scripts/update_lock.py

```diff
@@ -0,0 +1,206 @@
+OUTPUT_DEV_LOCK_PATH = REPO_ROOT_DIR / "requirements-dev.txt"
+
+
+def exclude_from_dependency_list(*, package_name: str, dependencies: list) -> list:
+    """Exclude the specified package from the provided dependency list."""
+
+    return [
+        dependency
+        for dependency in dependencies
+        if not dependency.startswith(package_name)
+    ]
+
+
+def remove_self_dependencies(pyproject: dict) -> dict:
+    """Filter out self dependencies (dependencies of the package on it self) from the
+    dependencies and optional-dependencies in the provided pyproject metadata."""
+
+    if not "project" in pyproject:
+        return pyproject
+
+    modified_pyproject = deepcopy(pyproject)
+
+    project_metadata = modified_pyproject["project"]
+
+    package_name = project_metadata.get("name")
+
+    if not package_name:
+        raise ValueError("The provided project metadata does not contain a name.")
+
+    if "dependencies" in project_metadata:
+        project_metadata["dependencies"] = exclude_from_dependency_list(
+            package_name=package_name, dependencies=project_metadata["dependencies"]
+        )
+
+    if "optional-dependencies" in project_metadata:
+        for group in project_metadata["optional-dependencies"]:
+            project_metadata["optional-dependencies"][
+                group
+            ] = exclude_from_dependency_list(
+                package_name=package_name,
+                dependencies=project_metadata["optional-dependencies"][group],
+            )
+
+    return modified_pyproject
+
+
+def fix_temp_dir_comments(file_path: Path):
+    """Fix the temp_dir comments so they don't cause a noisy diff
+
+    This will leave the top compile message intact as a point of sanity to verify that
+    the requirements are indeed being generated if nothing else changes.
+    """
+
+    with open(file_path, "r", encoding="utf-8") as file:
+        lines = file.readlines()
+
+    with open(file_path, "w", encoding="utf-8") as file:
+        for line in lines:
+            # Remove random temp directory name
+            line = re.sub(
+                r"\([^\)\(]*?pyproject\.toml\)",
+                "(pyproject.toml)",
+                line,
+            )
+            file.write(line)
+
+
+def compile_lock_file(
+    sources: list[Path],
+    output: Path,
+    upgrade: bool,
+    extras: bool,
+) -> None:
+    """From the specified sources compile a lock file using pip-compile from pip-tools
+    and write it to the specified output location.
+    """
+
+    print(f"Updating file at '{output}'...")
+
+    command = [
+        "pip-compile",
+        "--rebuild",
+        "--generate-hashes",
+        "--annotate",
+    ]
+
+    if upgrade:
+        command.append("--upgrade")
+
+    if extras:
+        command.append("--all-extras")
+
+    command += [
+        "--output-file",
+        str(output.absolute()),
+    ] + [str(source.absolute()) for source in sources]
+
+    with subprocess.Popen(
```

Comment:
Consider using subprocess.run instead of Popen.

Reply:
I'm not very familiar with the subprocess package, so this is a nice suggestion. Thanks